PDF backend can now do alpha on hatches.

mnmelo opened the pull request:
## PR Summary
Studied a bit of .PDF and I got this working. Turns out the transparency graphics states (`ExtGState`) weren't being taught to the hatch pattern writing routine.

This completes the fix of hatch alpha on SVG and PDF (#16883).

Furthermore, I noticed, after the previous PR (#16890), that there was a bit of mixed behavior between 2 cases: 1. when a Patch edgecolor (and therefore its hatch_color) has alpha; or 2. when the entire Patch has alpha. Case 1 is what I had tried to fix in #16890; however, the test I was relying on was of case 2.

Case 2 is more logically and elegantly rendered by doing edge+hatch full opacity and then within an object with alpha. Digging a bit deeper I found I could implement this case 2 behavior by extending the `forced_alpha` concept to the SVG and PDF renderers: if the parent Patch or Collection has a global alpha, ignore the ones of the elements.

I updated two `test_artist.py` tests to better cover these cases (no added baselines images). The following images showcase the results of the SVG and PDF `hatching` test, after conversion to PNG. The rightmost, orange-tinted shapes are the ones where alpha was set; on the top diamond only on the edgecolor, on the bottom one on the entire `Rectangle`:

_SVG_
![hatching_ svg](https://user-images.githubusercontent.com/11721337/78582850-d6c02700-782d-11ea-9bd0-011045531968.png)
Notice in the top-right shapes how the edge and face/hatch overlap. This no longer occurs if the alpha is set on the entire object (but then the fill is also affected).

_PDF_
![hatching_ pdf](https://user-images.githubusercontent.com/11721337/78582896-ef304180-782d-11ea-82b7-6ebbc9c5f13d.png)
Notice again in the top-right shape that PDF is cleaner in handling the edge overlap with the face and hatch of the same Patch. It's rendered as a group that is internally opaque (the edge completely occludes the fill and hatch) but still transparent at the edges (I guess this is called a knockout group? I didn't add anything specific to this end). I'm not sure SVG could allow for this but would be cool/cleaner if it did.

Feedback appreciated.

## PR Checklist

- [x] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

mnmelo commented on lib/matplotlib/tests/test_artist.py:
```diff
@@ -140,30 +148,50 @@ def test_cull_markers():
+                                               hatch='/', lw=4)
     ax.add_collection(rect2)
 
     # Ensure edge color is not applied to hatching.
```
I don't get why this comment is here: as it was and as it is now the edge color _does_ get applied to the hatching.

tacaswell replied:
That comment should be removed.  Doing a bit of git-blame digging and not removing that was an error on my part in https://github.com/matplotlib/matplotlib/pull/7976 (which restored hatches folliowing the edges see for bug report: https://github.com/matplotlib/matplotlib/issues/7901).  The hatches did not follow edges for 1 minor release via https://github.com/matplotlib/matplotlib/pull/7415 which is where the comment came in.

QuLogic commented:
If the .png changed too (and is intended!), then that needs a what's new as well.

mnmelo commented:
The .png was already behaving as intended (see #16883).

I'll update the PR soon with a fix of all the images that got affected. As I can tell at a glance, they're the ones involving PDF edge opacity, which apparently wasn't being handled properly (but perhaps I overlooked something).

mnmelo commented:
Tests should now pass.

Fixed a bug with `forced_alpha` I had introduced in the PDF backend. Besides that, checked and updated the four remaining failing images. These were:

`test_axes.py::test_contour_hatching`, `test_backend_svg.py::fill_black_with_alpha` and `test_backend_svg.py::patch_alpha_override`: These are all SVG where the previously wrong whole-patch alpha behavior is evident: by being applied to edge and fill separately their overlap becomes more opaque and yields a visible outline. This is now corrected as an outline-less patch. Example of `fill_black_with_alpha`:

_Before_
![fill_black_with_alpha-expected_svg](https://user-images.githubusercontent.com/11721337/78729125-22a7c480-7931-11ea-9a31-29d4ead6240a.png)
_Now_
![fill_black_with_alpha_svg](https://user-images.githubusercontent.com/11721337/78729134-276c7880-7931-11ea-9648-adf703fe3319.png)

And finally a PNG `test_mplot3d.py::poly3dcollection_alpha` now reflects the consolidated application of alpha to collections (more correctly, in my opinion, as edges were being rendered fully opaque).

tacaswell commented on lib/matplotlib/backends/backend_pdf.py:
```diff
@@ -1288,14 +1288,18 @@ def _write_soft_mask_groups(self):
             self.endStream()
 
-    def hatchPattern(self, hatch_style):
+    def hatchPattern(self, hatch_style, forced_alpha):
```
Does it make sense to give `forced_alpha` a default value of `False`?  That would preserve the API and behavior for anyone who happens to be using thins directly (which I hope there is not, but we have a lot of users who do a lot of very clever things).

tacaswell replied:
on the other hand I am the last person to have touched this to add `hatch_style`, did not do this and we had no complaints so we can probably ignore this comment.

jkseppan replied:
Maybe just add a warning in the release notes?

tacaswell commented on lib/matplotlib/collections.py:
```diff
@@ -299,6 +299,7 @@ def draw(self, renderer):
         gc = renderer.new_gc()
         self._set_gc_clip(gc)
         gc.set_snap(self.get_snap())
+        gc.set_alpha(self._alpha)
```
It is a bigger change, but I think that this means we can also remove a whole bunch of the complexity of managing the edge/face colors on collections.  If you look at `set_alpha` on collections you see we call `super().set_alpha()` and then call the private versions of `_set_facecolor` and `_set_edgecolor` with stashed versions of the original colors and then forces the alpha on them.  I _think_ that can go away.

mnmelo replied:
Well, stood out when looking at the collections code is that alpha is a second-class citizen, attribute-wise: you can give different face/edge colors to the collected elements, but the alpha always gets replaced by the collections' alpha. That made it easier to deem`gc.set_alpha(self._alpha)` safe. It does seem inconsistent, though. What should the goal be here? Simplicity or consistency?

jkseppan replied:
My suggestion is to make that a separate PR if it is deemed the correct thing to do. There is a lot of complexity there, some of which may be for a reason.

tacaswell commented:
Thanks for this, it is really good work! I am nominally :+1: on it going in, but would like to give @jkseppan a chance to weigh in (as the primary author of the pdf backend). 

I am not sure that I understand why you added the extra box to the clipping though.

mnmelo commented:
Why I added a backdrop rectangle to the clipping test? I first did it debug-wise to be sure there was alpha on the fill. Then thought it was also good for the clipping test itself, to check that it really is a clip, not a white overlay, so I let it stay. But I have no strong feeling either way; shall I remove it?

I still have to fix some linting, and apparently the Travis build is failing on macOS at the `bbox_inches_tight_raster` test... Any hints as to why? The macOS Azure builds succeeded.

The Windows Azure builds are also failing, because of `PIL.UnidentifiedImageError: cannot identify image file 'D:\\a\\1\\s\\result_images\\test_compare_images\\basn3p02-expected.png'`. This doesn't seem related to my changes.

Finally, what is your policy on squashing these final small-fix commits?

QuLogic commented:
The macOS failure is a known issue.

Squashing is up to the contributor, though it's preferred if you've changed test images multiple times.

mnmelo commented:
Good to know about the Mac failures. As to the Windows failures, I don't suppose there's a way of pulling the failing images out of the CI VM. I don't work on WIndows...

jkseppan commented on doc/users/next_whats_new/hatch-alpha.rst:
```diff
@@ -0,0 +1,6 @@
+
+The SVG and PDF backends now respect the hatch stroke alpha. Useful applications
+are, among others, semi-transparent hatches as a subtle way to differentiate
+columns in bar plots.
```
This could be a nice demo in the gallery, or perhaps another subplot in https://matplotlib.org/3.2.0/gallery/lines_bars_and_markers/filled_step.html

mnmelo replied:
Cool idea. As a side note, because hatch color/alpha is linked to the Patch's edge's, a simple plot with hatch alpha will usually not look so good because the translucent edge will half-overlap with the fill. One can plot edge-less with `lw=0`, but to have an opaque outline one must replot a full opacity, edge-only plot on top of the first.
It's a bit hackish. An independent hatch style API (#7059) would go a long way to simplifying this.

mnmelo commented:
Ok, I'll be splitting the collection alpha into a separate PR, and try to add/update a gallery example as suggested.

QuLogic commented:
Is this waiting for the separate PR, or vice versa?

I wonder if #10035 is a similar sort of bug for markers, as the Patch edgecolor handling.

mnmelo commented:
Both; this should not be merged just yet. I haven't had the time to spin-off a collections-targeted PR.

As to #10035, this might very well be connected. I'll take a look. After a quick read it seems SVG supports 'knock-out' or is going to support it (not sure about the stability of that). This would further homogeneize SVG and PDF behavior. A discussion then would be on what is the intended way of displaying non-opaque edges over their fills (I vote for the cleaner-looking knockout compositing).

oscargus commented:
When this is continued, maybe this issue can be considered as well? https://github.com/matplotlib/matplotlib/issues/12367 (Seems like it is most likely fixed by this PR.)
